backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt

```diff
@@ -627,25 +627,25 @@ internal class FunctionGenerationContext(val function: LLVMValueRef,
             LLVMAddClause(landingpad, kotlinExceptionRtti.llvm)
             LLVMAddClause(landingpad, LLVMConstNull(kInt8Ptr))
 
-            val fatalForeignExceptionBlock = basicBlock("fatalForeignException", position()?.start)
+            val forwardNativeExceptionBlock = basicBlock("forwardNativeException", position()?.start)
             val forwardKotlinExceptionBlock = basicBlock("forwardKotlinException", position()?.start)
 
             val isKotlinException = icmpEq(
                     extractValue(landingpad, 1),
                     call(context.llvm.llvmEhTypeidFor, listOf(kotlinExceptionRtti.llvm))
             )
 
-            condBr(isKotlinException, forwardKotlinExceptionBlock, fatalForeignExceptionBlock)
+            condBr(isKotlinException, forwardKotlinExceptionBlock, forwardNativeExceptionBlock)
 
             appendingTo(forwardKotlinExceptionBlock) {
                 // Rethrow Kotlin exception to real handler.
                 codeContext.genThrow(extractKotlinException(landingpad))
             }
 
-            appendingTo(fatalForeignExceptionBlock) {
-                val exceptionRecord = extractValue(landingpad, 0)
-                call(context.llvm.cxaBeginCatchFunction, listOf(exceptionRecord))
-                terminate()
+            appendingTo(forwardNativeExceptionBlock) {
+                val exception = createForeignException(landingpad, codeContext.exceptionHandler)
+                codeContext.genThrow(exception)
+                unreachable()
```

Comment:
`unreachable` shouldn't be required here, `genThrow` is expected to take care of that.